tools/pipelines/templates/include-publish-npm-package-steps.yml

```diff
@@ -55,112 +55,133 @@ steps:
       echo Tag: $tag
       cp .npmrc ~/.npmrc
       maximumRetryIfNetworkError=3
+
+      # Note: the package names here are "processed" the way 'npm pack' does (if the package is scoped, remove
+      # the leading @ and replace / with -).
+      # When getting them from a txt file that's the way we wrote them to it.
+      # When getting them from the output of 'ls' that's also how they are because the files are the 'npm pack'ed
+      # packages but we need to strip the version number and the extension for consistency, so later we can assume
+      # that these are just the package names, as we need them like that for some things.
       if [[ -f $(Pipeline.Workspace)/pack/packagePublishOrder-${{ parameters.feedKind }}.txt ]]; then
-        packages=$(cat $(Pipeline.Workspace)/pack/packagePublishOrder-${{ parameters.feedKind }}.txt)
-        sorted=true
+        processedPackageNames=$(cat $(Pipeline.Workspace)/pack/packagePublishOrder-${{ parameters.feedKind }}.txt)
+        fromPackageList=true
       else
-        packages=*.tgz
-        sorted=false
+        # The sed expression removes from the first dash followed by a number, to the file extension.
+        # This assumes we don't have packages whose names have a dash followed by a number.
+        # See comment above for more details.
+        processedPackageNames=$(ls *.tgz | sed 's/-[0-9].*\.tgz//g')
       fi
-      for packageName in $packages
+      for processedPackageName in $processedPackageNames
       do
-        if [[ $sorted == true ]]; then
-          f=$(find -name "${packageName}-[0-9]*.tgz")
-        else
-          f=$packageName
+        # The filenames are the processed package name followed by a dash, then the version number, then the extension.
+        # This glob pattern needs to account for variations in our version number schemes, and the fact that
+        # some package names might be substrings of others, and does it by looking for a filename where the
+        # package name is immediately followed by a dash and a number (such that, for example, when looking for
+        # @fluidframework/myPackage, we hit @fluidframework/myPackage-2.0.0 and not a potential
+        # @fluidframework/myPackage-legacy-2.0.0).
+
+        f=$(find -name "${processedPackageName}-[0-9]*.tgz")
+
+        if [[ $f == "" ]]; then
+          # TODO: this seems like it should never happen and should be a fatal error?
+          # Don't want to be too aggressive with this change, so leaving it as a warning for now and keeping the
+          # current behavior that just skips this case.
+          echo "##[warning]Could not find a file for package $processedPackageName"
+          continue
         fi
-        if [[ $f != "" ]]; then
```

Comment:
Replaced this with the opposite check a few lines above to log a warning and skip the current iteration (current `processedPackageName`), so everything under here didn't need extra nesting.